kernel/reduction.ml

```diff
@@ -42,20 +42,64 @@ let rec zip_lists l1 l2 lst =
 
 (* State *)
 
-type env = term Lazy.t LList.t
 
 (* A state {ctx; term; stack} is the state of an abstract machine that
 represents a term where [ctx] is a ctx that contains the free variables
 of [term] and [stack] represents the terms that [term] is applied to. *)
 type state = {
-  ctx:env;        (*context*)
-  term : term;    (*term to reduce*)
+  ctx   : env;        (*context*)
+  term  : term;    (*term to reduce*)
   stack : stack;  (*stack*)
+  reduc : (bool * state) ref;
```

Comment:
Out of curiosity, why do you use here, (1) a tuple and (2) a reference ? Both add an extra level of indirection, while it seems two mutable fields would work as well. The fewer indirection might provide some performance boost since it would probably help the locality of data (and ultimately cpu caching).

For reference see [here](https://github.com/Gbury/ocaml-memgraph/blob/master/examples/dk.ml) and [here](https://github.com/Gbury/ocaml-memgraph/blob/master/examples/dk.svg) for a graph of the memory layout of both options.

Reply:
I agree, there is at least one unnecessary level of indirection here. This was for the sake of simplicity, this PR is not meant to be merged just yet, I'm just messing with the reduction machine in a "quick and dirty" way.

For the record, I believe there may be a use for one "level of indirection" to allow sharing of a common mutable "reference to reduct" between multiple distinct states.

However after talking with Bruno it seems the best way to handle all this sharing would actually be a union-find data structure.